sdks/cpp/common/examples/full_service/full_service.cpp

```diff
@@ -156,24 +163,68 @@ std::shared_ptr<grpc::ServerCredentials> getServerCredentials() {
 //
 // right now, it just tests that a token exists, decodes it and prints it out
 // much work required to actually validate the token
-void authorize(grpc::ServerContext *context /*catena::Param &p*/) {
-    if (absl::GetFlag(FLAGS_authz)) {
-        auto authz = context->client_metadata();
-        auto it = authz.find("authorization");
-        if (it == authz.end()) {
-            BAD_STATUS("No authorization token found", catena::StatusCode::PERMISSION_DENIED);
+class RPCInterceptor : public Interceptor {
+  public:
+    RPCInterceptor(DeviceModel &dm) : dm_(dm) {}
+
+    void Intercept(InterceptorBatchMethods* methods) override {
+        if (methods->QueryInterceptionHookPoint(
+            InterceptionHookPoints::POST_RECV_INITIAL_METADATA)) {
+            std::cout << "Got a new streaming RPC" << std::endl;
+
+            if (absl::GetFlag(FLAGS_authz)) {
+                auto authz = methods->GetRecvInitialMetadata();
+                auto it = authz->find("authorization");
+                if (it == authz->end()) {
+                    BAD_STATUS("No authorization token found", catena::StatusCode::PERMISSION_DENIED);
```

Comment:
we shouldn't be using exceptions to do what looks more like flow control. The logic here should be to (somehow) return the PERMISSION_DENIED to the client, and not pass the RPC to the next layer (i.e. don't call the Proceed method) instead of throwing exceptions.